### PR TITLE
CAD-2772 add cardano-tx-generator

### PR DESCRIPTION
**Pull request by MarcFontaine**
* Add tx-generator package.
  (copy https://github.com/input-output-hk/cardano-benchmarking/tree/master/cardano-tx-generator)
* Update default.nix and cabal.project.

### CODE REVIEW

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
Nice! However, I don't think this belong in cardano-node.

**Comment by erikd** [comment]
Unless there is an incredibly good reason to do move it to the `node` repo it should stay in `iohk-monitoring` .

**Comment by MarcFontaine** [comment]
The background of this PR is the plan to run benchmarks for a cluster of nodes as part of the `cardano-node` CI
(either nightly or even on a per-PR basis). The nix-setup and the scripts for these benchmarks in "some non trivial way" depend on the `tx-generator` being part of the node repository. The benchmarking team will write down a more detailed explanation in an external document. (I'll post the link.)

**Comment by erikd** [comment]
The fact that these are needed to run the benchmarks does still not mean this should be in the node repo.

**Review by erikd — CHANGES_REQUESTED** [review]
This does not belong in the `node` repo.

**Comment by MarcFontaine** [comment]
@erikd @Jimbo4350 could you please have a look at this PR again.

File: cardano-node/cardano-node.cabal
Comment (deepfire) [inline cardano-node/cardano-node.cabal:4]: @MarcFontaine, this has slipped by.

File: default.nix
Comment (deepfire) [inline default.nix:30]: @MarcFontaine, this has slipped by.

File: release.nix
Comment (deepfire) [inline release.nix:129]: @MarcFontaine, also slipped by.

File: release.nix
Comment (deepfire) [inline release.nix:130]: @MarcFontaine, re `musl` issues -- try to change this line to 
`[ "tx-generator" "components" "exes" ]`

File: release.nix
Comment (MarcFontaine) [reply release.nix:129]: This is unclear.
Did I add that line by mistake ?

Comment (deepfire) [reply release.nix]: This is what this PR adds to the baseline, indeed -- and it shouldn't, since this line is completely unrelated to the generator.

**Comment by deepfire** [comment]
Ok, the build had a propagated failure, restarted.

File: release.nix
Comment (deepfire) [reply release.nix]: Ok, we found the right incantation.

**Review by deepfire — APPROVED** [review]
LGTM!

**Comment by angerman** [comment]
The fix in 46a1ead is likely due to some bug in haskell.nix, specifically in
https://github.com/input-output-hk/haskell.nix/blob/621f52e29b1d9dce372ef7aee747bcddeb915c8a/builder/comp-builder.nix#L45-L48

once haskell.nix is fixed these kinds of shenanigans should not be needed anymore.

**Review by erikd — DISMISSED** [review]
There should not be any commits comtaining text like "rebase to master".

Commits messages like this contain zero useful information. I cannot see why this is not a single commit.

File: bench/tx-generator/src/Cardano/Benchmarking/CliArgsScript.hs
Comment (erikd) [inline bench/tx-generator/src/Cardano/Benchmarking/CliArgsScript.hs:65]: Pretty sure the rest of the node (on `master`) already supports `Alonzo`, so we should fix this.

Comment (newhoggy) [reply bench/tx-generator/src/Cardano/Benchmarking/CliArgsScript.hs:65]: Yeah, or some explanation of why we don't support alonzo era and what impact that might have.

Comment (MarcFontaine) [reply bench/tx-generator/src/Cardano/Benchmarking/CliArgsScript.hs:65]: We do not want to lose all the history of the tx-generator. This PR is (feature wise) the latest version of tx-genrator
when we stopped development in the old repo.
After it is merged to `cardano-node` we plan to remove some of the features and `CliArgsScript.hs` will be removed all together.

Comment (erikd) [reply bench/tx-generator/src/Cardano/Benchmarking/CliArgsScript.hs:65]: > We do not want to lose all the history of the tx-generator. 

What history?

The `master` branch of the node currently supports `Alonzo`. If this PR is merged it will almost certainly *break* it .

> After it is merged to cardano-node we plan to remove some of the features and CliArgsScript.hs will be removed all together.

No, you need to fix it first, then it can be merged.

File: bench/tx-generator/README.md
Comment (Jimbo4350) [inline bench/tx-generator/README.md:1]: Can we add an introduction as to how to use the transaction generator and list the team responsible for this library?

**Review by Jimbo4350 — APPROVED** [review]
Minor changes.

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
Two minor changes.

File: bench/tx-generator/src/Cardano/Benchmarking/CliArgsScript.hs
Comment (deepfire) [reply bench/tx-generator/src/Cardano/Benchmarking/CliArgsScript.hs:65]: @erikd, that support is being worked on in an follow-up branch, and it's a non-trivial bit.

Comment (deepfire) [reply bench/tx-generator/src/Cardano/Benchmarking/CliArgsScript.hs:65]: Moreover, you are looking at the CLI-based configuration for the generator, which is only needed for a very short interim period, and will be removed quite soon.

Comment (deepfire) [reply bench/tx-generator/src/Cardano/Benchmarking/CliArgsScript.hs:65]: And yet still, the generator can be used with Alonzo clusters -- it's just that it can't generate Alonzo-specific transactions yet.

Comment (erikd) [reply bench/tx-generator/src/Cardano/Benchmarking/CliArgsScript.hs:65]: So when are the commits going to be squashed?

**Comment by Jimbo4350** [comment]
@deepfire @MarcFontaine Guys the coding style appears inconsistent. Can you spend a little time cleaning up the formatting?

File: bench/tx-generator/src/Cardano/Benchmarking/CliArgsScript.hs
Comment (MarcFontaine) [reply bench/tx-generator/src/Cardano/Benchmarking/CliArgsScript.hs:65]: Serge did a first squash Yesterday.
We will squash it again, after the reviews, when it is ready to merge.

Comment (erikd) [reply bench/tx-generator/src/Cardano/Benchmarking/CliArgsScript.hs:65]: I would must rather review *after* the squash rather than before.

**Comment by MarcFontaine** [comment]
OK, I changed the formatting of the imports..

**Review by Jimbo4350 — APPROVED** [review]
Be sure to get @erikd's approval.

**Comment by Jimbo4350** [comment]
> OK, I changed the formatting of the imports..

The code is still somewhat inconsistent regarding indentation.

**Pull request merged**
